primo/data_parser/metric_data.py

```diff
@@ -0,0 +1,602 @@
+
+    def delete_submetric(self, name):
+        """
+        Deletes an existing submetric
+
+        Parameters
+        ----------
+        name : str
+            Submetric `name`/`full_name`
+        """
+        _extended_metrics = self._get_all_metrics_extended
+        obj = _extended_metrics.get(name, None)
+
+        # Raise an error if the metric does not exist
+        if obj is None:
+            raise AttributeError(f"Submetric {name} does not exist.")
+
+        parent_obj = obj.parent_metric
+        parent_obj.submetrics.pop(obj.name)
+        delattr(self, obj.name)
+
+        # If all submetrics have been removed, then delete the attribute
+        if len(parent_obj.submetrics) == 0:
+            delattr(parent_obj, "submetrics")
+
+    def set_value(self, primary_metrics, submetrics=None, check_validity=True):
+        """
+        Sets/updates the weights of all metrics/submetrics.
+
+        Parameters
+        ----------
+        primary_metrics : dict
+            Dictionary containing the weights of primary metrics.
+            key must be the primary metric's `name` and the value
+            must be its weight. Submetrics can also be included in this
+            dictionary.
+
+        submetrics : dict(dict), default = None
+            Dictionary of dictionaries to update the weights of submetrics.
+            This is optional, and it is useful to avoid confusion. The data must
+            be of the form
+            {primary_metric_name: {"submetric_1": weight_1, "submetric_2": weight_2}}
+
+        check_validity : bool, default=True
+            Checks if the weights are valid i.e., add up to 100
+        """
+        # Having a mutable object as a default is not recommended.
+        # Apparently, this is the recommended way to set a mutable object
+        # as the default.
+        if submetrics is None:
+            submetrics = {}
+
+        _extended_metrics = self._get_all_metrics_extended
+
+        # Add submetrics to primary metrics
+        for key in submetrics:
+            primary_metrics.update(submetrics[key])
+
+        for key, obj in _extended_metrics.items():
+            if key in primary_metrics:
+                obj.value = primary_metrics.pop(key)
+
+        # If there are any unused elements, raise an error.
+        if len(primary_metrics) > 0:
+            raise KeyError(
+                f"Metrics/submetrics {[*primary_metrics]} are not recognized/registered."
+            )
+
+        # Check validity of the data
+        if check_validity:
+            self.check_validity()
+
+    def check_validity(self):
+        """
+        Checks the validity of the weight values.
+        Returns None if the weights are valid.
+        Raises ValueError if the weights are not valid.
+        """
+        primary_metric_sum = sum(
+            obj.weight for obj in self.get_primary_metrics.values()
+        )
+
+        if abs(primary_metric_sum - 100) > 0.01:
```

Comment:
Suggest using np.isclose as it helps avoid introducing a constant 0.01. Also avoids confusion on why line 478 uses 0.001 as the tolerance instead of 0.01